### PR TITLE
Feature/LOC

### PR DESCRIPTION
**Pull request by idodeclare**
Hello,

Please consider for integration this patch to count per-file physical lines-of-code ([LOC](http://csse.usc.edu/TECHRPTS/2007/usc-csse-2007-737/usc-csse-2007-737.pdf);
 [Wiki](https://en.wikipedia.org/wiki/Source_lines_of_code)) and to store LOC and total number of lines.

`DirectoryListing` is updated to [present the counts](https://1drv.ms/i/s!Anl22C9Mx4VFgRqzmGmJ3Lzszyt6) while browsing.

Thank you.

### CODE REVIEW

**Comment by idodeclare** [comment]
I should note that this resolved issue #591.

**Comment by tulinkry** [comment]
I would personally prefer if all the methods mentioning LOC would have it expanded to LinesOfCode.

File: src/org/opensolaris/opengrok/index/IndexDatabase.java [hunk 1]
```diff
@@ -272,7 +274,7 @@ public void run() {
 
     @SuppressWarnings("PMD.CollapsibleIfStatements")
     private void initialize() throws IOException {
-        synchronized (this) {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/index/IndexDatabase.java]: does this make any difference?

Comment (idodeclare) [reply src/org/opensolaris/opengrok/index/IndexDatabase.java]: Just that synchronizing on public objects is suspect as it breaks encapsulation — so I revised it.

File: src/org/opensolaris/opengrok/index/IndexDatabase.java [hunk 2]
```diff
@@ -805,7 +805,7 @@ private boolean accept(File file) {
         return !RuntimeEnvironment.getInstance().isIndexVersionedFilesOnly();
     }
 
-    boolean accept(File parent, File file) {
+    private boolean accept(File parent, File file) {
```
Comment (tulinkry) [inline src/org/opensolaris/opengrok/index/IndexDatabase.java]: is this relevant? or just a cleanup?

**Comment by idodeclare** [comment]
> I would personally prefer if all the methods mentioning LOC would have it expanded to LinesOfCode.

I was motivated to have it quite terse for the methods that are used in analyzers, since I had to edit hundreds of lines in them. Moreover since LOC is almost orthogonal to the bulk of logic in the xrefers, I wanted a tiny function name — `chkLOC()` — rather than seeing `checkLinesOfCode()` over and over (and also having to reformat possibly hundreds of newly too-long lines).

I do think "LOC" is well-known enough acronym in source code analysis; if analyzers are ever updated to support logical lines-of-code, then "LLOC" also will be well-known.

File: src/org/opensolaris/opengrok/index/IndexDatabase.java (hunk 2 above)
Comment (idodeclare) [reply src/org/opensolaris/opengrok/index/IndexDatabase.java]: Just a cleanup

**Comment by ChristopheBordieu** [comment]
Excellent feature!!!
Will all the total lines and slocs be aggregated at repository root level ?

**Comment by vladak** [comment]
@idodeclare pls use 'fixes #591' in one the commit comments then.

**Comment by idodeclare** [comment]
> Will all the total lines and slocs be aggregated at repository root level ?

That is certainly feasible in the future. Right now, only files have document data in the Lucene index. To store aggregated num-lines and LOC, it will be necessary to store Lucene data for directories — but without conflicting with the existing queries that all assume only file-level data.

One straight-forward idea would be to add Lucene "documents" for directories with wholly-separate fields so that these new documents never appear in current queries. Of course it would also be necessary to write an aggregator that runs at the appropriate stages.

**Comment by vladak** [comment]
This also brings a question about how to use these counts for restricting
searches.

Dne 26. 1. 2018 5:10 odp. napsal uživatel "C Fraire" <
notifications@github.com>:

**Comment by idodeclare** [comment]
> This also brings a question about how to use these counts for restricting
searches.

Yes, indeed.

Relatedly, I did not update to show the counts in search results, as I did not have any idea how best to present them.

**Comment by tarzanek** [comment]
I am merging this as is, thank you Chris, it looks good

**Comment by idodeclare** [comment]
Thanks, @tarzanek !

**Comment by vladak** [comment]
I wonder how this will cope with incremental indexing. Will pre-existing files have LOC set to 0 ? Or is full reindex required ?

**Comment by idodeclare** [comment]
Pre-existing documents will have blank LOC/#Lines — the same as for files which do not get an xref (e.g. `Ignored/FileAnalyzer`) or for files that do not have a counting xref (e.g. `UuencodeAnalyzer`).

Incrementally re-indexed files can get LOC/#Lines. Full re-index is required for completeness.

**Pull request merged**